File: apps/web/src/types/jobs.ts
export type JobType = "WORKFLOW_RUN";

export type JobStatus =
  "QUEUED" | "RUNNING" | "SUCCEEDED" | "FAILED" | "CANCEL_REQUESTED" | "CANCELLED";

export type JobErrorCategory =
  "VALIDATION" | "DATA" | "INFRASTRUCTURE" | "WORKER" | "INTERNAL";

export interface JobProgress {
  current: number;
  total: number;
  label: string;
}

export interface JobAttempt {
  id: string;
  job_id: string;
  attempt_number: number;
  status: JobStatus;
  worker_id: string | null;
  started_at: string;
  completed_at: string | null;
  duration_ms: number | null;
  error_code: string | null;
  error_message_safe: string | null;
  error_category: JobErrorCategory | null;
}

export interface JobFailure {
  code: string | null;
  message: string | null;
  category: JobErrorCategory | null;
  retryable: boolean;
}

export interface JobSummary {
  id: string;
  job_type: JobType;
  status: JobStatus;
  workflow_run_id: string;
  workflow_id: string | null;
  workflow_name: string | null;
  dataset_id: string | null;
  dataset_name: string | null;
  input_version_id: string | null;
  input_version_number: number | null;
  output_version_id: string | null;
  output_version_number: number | null;
  queue_name: string;
  attempt_count: number;
  max_attempts: number;
  progress: JobProgress;
  current_step_position: number | null;
  current_operation_code: string | null;
  current_activity: string | null;
  queued_at: string;
  started_at: string | null;
  completed_at: string | null;
  heartbeat_at: string | null;
  queue_ms: number | null;
  execution_ms: number | null;
  total_ms: number | null;
  error_code: string | null;
  error_message_safe: string | null;
  error_category: JobErrorCategory | null;
  retryable: boolean;
  created_at: string;
  updated_at: string;
}

export interface JobDetail extends JobSummary {
  request_id: string | null;
  cancel_requested_at: string | null;
  cancelled_at: string | null;
  attempts: JobAttempt[];
  step_runs: import("@/types/workflows").WorkflowStepRun[];
  workflow_revision: number | null;
  workflow_run_status: string | null;
  quality_before: number | null;
  quality_after: number | null;
  rows_before: number | null;
  rows_after: number | null;
}

export interface JobListResponse {
  items: JobSummary[];
  page: number;
  page_size: number;
  total: number;
}

export interface JobFilters {
  status?: JobStatus;
  job_type?: JobType;
  workflow_id?: string;
  dataset_id?: string;
  q?: string;
  page?: number;
  page_size?: number;
}

export interface QueueHealth {
  status: "ready" | "not_configured" | "unavailable";
  backend: string;
  queued_count: number | null;
  name: string;
}

export interface WorkerHealth {
  status: "available" | "unavailable";
  available_count: number;
  last_seen_at: string | null;
}

export interface OperationsHealth {
  queue: QueueHealth;
  worker: WorkerHealth;
  jobs: {
    queued: number;
    running: number;
    failed: number;
    succeeded: number;
  };
}

export const TERMINAL_JOB_STATUSES: readonly JobStatus[] = [
  "SUCCEEDED",
  "FAILED",
  "CANCELLED",
];

export function isActiveJobStatus(status: JobStatus): boolean {
  return status === "QUEUED" || status === "RUNNING" || status === "CANCEL_REQUESTED";
}

File: apps/api/src/facilio/schemas/jobs.py
"""Job and operational health API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from facilio.schemas.workflows import WorkflowRunDetail, WorkflowStepRunData

JobType = Literal["WORKFLOW_RUN"]
JobStatus = Literal[
    "QUEUED",
    "RUNNING",
    "SUCCEEDED",
    "FAILED",
    "CANCEL_REQUESTED",
    "CANCELLED",
]
JobErrorCategory = Literal["VALIDATION", "DATA", "INFRASTRUCTURE", "WORKER", "INTERNAL"]


class JobAttemptData(BaseModel):
    id: UUID
    job_id: UUID
    attempt_number: int
    status: JobStatus
    worker_id: str | None = None
    started_at: datetime
    completed_at: datetime | None = None
    duration_ms: int | None = None
    error_code: str | None = None
    error_message_safe: str | None = None
    error_category: JobErrorCategory | None = None


class JobProgressData(BaseModel):
    current: int
    total: int
    label: str


class JobSummary(BaseModel):
    id: UUID
    job_type: JobType
    status: JobStatus
    workflow_run_id: UUID
    workflow_id: UUID | None = None
    workflow_name: str | None = None
    dataset_id: UUID | None = None
    dataset_name: str | None = None
    input_version_id: UUID | None = None
    input_version_number: int | None = None
    output_version_id: UUID | None = None
    output_version_number: int | None = None
    queue_name: str
    attempt_count: int
    max_attempts: int
    progress: JobProgressData
    current_step_position: int | None = None
    current_operation_code: str | None = None
    current_activity: str | None = None
    queued_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    heartbeat_at: datetime | None = None
    queue_ms: int | None = None
    execution_ms: int | None = None
    total_ms: int | None = None
    error_code: str | None = None
    error_message_safe: str | None = None
    error_category: JobErrorCategory | None = None
    retryable: bool = False
    created_at: datetime
    updated_at: datetime


class JobDetail(JobSummary):
    request_id: str | None = None
    cancel_requested_at: datetime | None = None
    cancelled_at: datetime | None = None
    attempts: list[JobAttemptData] = Field(default_factory=list)
    step_runs: list[WorkflowStepRunData] = Field(default_factory=list)
    workflow_revision: int | None = None
    workflow_run_status: str | None = None
    quality_before: float | None = None
    quality_after: float | None = None
    rows_before: int | None = None
    rows_after: int | None = None


class JobListData(BaseModel):
    items: list[JobSummary]
    page: int
    page_size: int
    total: int


class WorkflowRunAccepted(BaseModel):
    workflow_run: WorkflowRunDetail
    job: JobSummary


class OperationsHealthData(BaseModel):
    queue: dict[str, Any]
    worker: dict[str, Any]
    jobs: dict[str, int]
